tests/testsuite/publish.rs

```diff
@@ -815,6 +784,85 @@ fn block_publish_no_registry() {
 [ERROR] some crates cannot be published.
 `foo` is marked as unpublishable
 ",
-        )
-        .run();
+        ).run();
+}
+
+#[test]
+fn publish_with_select_features() {
+    publish::setup();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            cargo-features = ["alternative-registries"]
+
+            [project]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+            license = "MIT"
+            description = "foo"
+            publish = []
+
+            [features]
+            required = []
+            optional = []
+        "#,
+        ).file(
+            "src/main.rs",
+            "#[cfg(not(required))]
+             compile_error!(\"This crate requires `required` feature!\");
+             fn main() {}",
+        ).build();
+
+    p.cargo("publish --registry alternative -Zunstable-options --features required")
```

Comment:
I'm not sure why these tests are set up this way with alternative registries.  You can publish to a local (fake) registry.  See `simple` for an example.

Reply:
Hey, so I finally got the chance to look at this again.

I think the reason why most of the other tests are using the alternative registry is so that they can omit options from the `Cargo.toml`, which is technically not valid (anymore?).

Without an alternative registry I'll need to provide an author which is fine by itself, just means that it would be the only test in the file to provide one.

I opted for "do it the same way all the other tests work" over trying to be clever.

Do you think that's okay? Otherwise I'll happily change the test and just add more fields to the manifest as are needed.

Reply:
The nearby tests that are using alternative registries are doing so to test the `publish` field which requires that feature.

I'm not sure I understand what you mean by the author field, since there is one in the test.  Here's the style I'm thinking of:

```diff
diff --git a/tests/testsuite/publish.rs b/tests/testsuite/publish.rs
index dca4e92c..7b3dce40 100644
--- a/tests/testsuite/publish.rs
+++ b/tests/testsuite/publish.rs
@@ -812,15 +812,12 @@ fn publish_with_select_features() {
         .file(
             "Cargo.toml",
             r#"
-            cargo-features = ["alternative-registries"]
-
             [project]
             name = "foo"
             version = "0.0.1"
             authors = []
             license = "MIT"
             description = "foo"
-            publish = []

             [features]
             required = []
@@ -829,20 +826,15 @@ fn publish_with_select_features() {
         )
         .file(
             "src/main.rs",
-            "#[cfg(not(required))]
+            "#[cfg(not(feature=\"required\"))]
              compile_error!(\"This crate requires `required` feature!\");
              fn main() {}",
         ).build();

-    p.cargo("publish --registry alternative -Zunstable-options --features required")
-        .masquerade_as_nightly_cargo()
-        .with_status(101)
-        .with_stderr(
-            "\
-[ERROR] some crates cannot be published.
-`foo` is marked as unpublishable
-",
-        ).run();
+    p.cargo("publish --features required --index")
+        .arg(publish::registry().to_string())
+        .with_stderr_contains("[UPLOADING] foo v0.0.1 ([CWD])")
+        .run();
 }

 #[test]
```